moveit_ros/moveit_servo/src/servo.cpp

```diff
@@ -135,6 +135,23 @@ Servo::Servo(const rclcpp::Node::SharedPtr& node, std::shared_ptr<const servo::P
     joint_name_to_index_maps_.insert(
         std::make_pair<std::string, JointNameToMoveGroupIndexMap>(std::string(sub_group_name), std::move(new_map)));
   }
+
+  const moveit::core::JointModelGroup* joint_model_group =
+      robot_state->getJointModelGroup(servo_params_.move_group_name);
+  const std::vector<std::string> joint_names = joint_model_group->getActiveJointModelNames();
+  const moveit::core::JointBoundsVector joint_bounds = joint_model_group->getActiveJointModelsBounds();
+  joint_acceleration_limit = -1.0;
+  for (const auto& bound : joint_bounds)
+  {
+    for (const auto& b1 : *bound)
+    {
+      double max_acceleration = servo_params_.scale.joint_acceleration *
+                                std::max(std::abs(b1.min_acceleration_), std::abs(b1.min_acceleration_));
+      joint_acceleration_limit =
+          joint_acceleration_limit < 0 ? max_acceleration : std::min(joint_acceleration_limit, max_acceleration);
+    }
+  }
```

Comment:
Few comments here:

First off, these acceleration joint bounds I don't think are correct for a couple reasons.
* You don't just want to take the lowest value of all the joints; rather, you want to look at the un-limited target command and determine the limiting joint such that scaling down meets all the joint limits.
* This means that this shouldn't be computed as a member variable of servo, but just  with local variables at every iteration.
* If the scale parameter is acting as a true scale factor value, we should disallow exceeding the joint limits, meaning the scale always needs to be in the range `(0..1]`.
* If this is not acting as a scale factor, but as an actual acceleration in rad/s^2, then similarly you want to scale down all the computed target accelerations and scale them down to the value provided... UNLESS they exceed the actual joint's limits.
* For better variable naming, see what is done for velocity scaling: https://github.com/ros-planning/moveit2/blob/main/moveit_ros/moveit_servo/src/utils/common.cpp#L324-L338

As such, I think this should be moved to a utility in a place similar to what I linked above, and ensure it's unit tested.

---

Secondly, what does it mean when the robot does not have any max accelerations provided? This again boils down to a similar question as above:

* If the scale parameter is acting as a true scale factor, then we need to check if the joint HAS acceleration bounds (there are variables/functions for this in the joint bounds), and if not then don't limit acceleration.
* If this is not acting as a scale factor, but as an actual acceleration in rad/s^2 ... then we just take the parameter as max acceleration and move on with scaling.

Reply:
> First off, these acceleration joint bounds 

Thank you for pointing this out. What I did was make a new function called `jointLimitAccelerationScalingFactor` which is like `jointLimitVelocityScalingFactor` except that it acts on acceleration instead of velocity.  Unfortunately, it is a lot of copied code. There are two ways to refactor that I can see. Have both methods call an internal method that 1) takes lambda that calculates the scaling factor for either velocity or acceleration or 2) takes a boolean flag and has an `if else` to either do velocity or acceleration.

> If the scale parameter is acting as a true scale factor value, we should disallow exceeding the joint limits, meaning the scale always needs to be in the range (0..1].    

We need to rethink scaling. I think a unitless factor makes more sense than an actual limit in m/s^2. I think we should have a separate parameter for scaling the dynamic joint limits. This would apply to both velocity and acceleration. None of the existing parameters actually scale these limits, it's either scaling the command or scaling the velocity to be a certain amount less than the limit. 

> Secondly, what does it mean when the robot does not have any max accelerations provided?

I think if the robot has no maximum acceleration, then we should either 1) not apply the acceleration scaling or 2) throw an exception. I prefer option 1), especially since the velocity scaling already works that way.

Reply:
I also added a test for `jointLimitAccelerationScalingFactor`

Reply:
I like option 1 for both of these.

I was recently told that having a boolean parameter for a function to do two different things means you should have 2 separate functions... or make it generic via things like templates/lambdas